# HG changeset patch
# User Julien Bect <jbect@users.sourceforge.net>
# Date 1546349668 -3600
#      Tue Jan 01 14:34:28 2019 +0100
# Branch stable
# Node ID 7a77eb39c7f0ad8b342a4c99060a8fd383a75c65
# Parent  b3157525cebfe83bdd79f6948429eb1030217c98
jit: configure: Check if llvm/Transforms/Scalar/GVN.h exists

* configure.ac: Check if llvm/Transforms/Scalar/GVN.h exists.
* libinterp/parse-tree/pt-jit.cc: If llvm/Transforms/Scalar/GVN.h exists,
  we need to include it in order to use llvm::createGVNPass.

diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -1431,16 +1431,17 @@ if test $ENABLE_JIT = yes; then
     AC_CHECK_HEADERS([llvm/Target/TargetData.h llvm/IR/DataLayout.h \
                       llvm/DataLayout.h], [have_llvm_data_h=yes; break])
     if test $have_llvm_data_h = no; then
       warn_llvm="Missing LLVM file TargetData.h.  JIT compiler is disabled."
     fi
 
     AC_CHECK_HEADERS([llvm/IR/Verifier.h])
     AC_CHECK_HEADERS([llvm/Analysis/BasicAliasAnalysis.h])
+    AC_CHECK_HEADERS([llvm/Transforms/Scalar/GVN.h])
 
     OCTAVE_LLVM_FUNCTION_ADDATTRIBUTE_API
     OCTAVE_LLVM_FUNCTION_ADDFNATTR_API
     OCTAVE_LLVM_CALLINST_ADDATTRIBUTE_API
     OCTAVE_LLVM_RAW_FD_OSTREAM_API
     OCTAVE_LLVM_LEGACY_PASSMANAGER_API
     OCTAVE_LLVM_IRBUILDER_API
     AC_LANG_POP(C++)
diff --git a/libinterp/parse-tree/pt-jit.cc b/libinterp/parse-tree/pt-jit.cc
--- a/libinterp/parse-tree/pt-jit.cc
+++ b/libinterp/parse-tree/pt-jit.cc
@@ -106,16 +106,23 @@ along with Octave; see the file COPYING.
 #  include <llvm/DataLayout.h>
 #else
 #  include <llvm/Target/TargetData.h>
 #endif
 
 #include <llvm/Transforms/IPO.h>
 #include <llvm/Transforms/Scalar.h>
 
+// Starting with LLVM 3.9.0, llvm::createGVNPass has
+// been moved to a new header file named GVN.h
+// (before that it was in llvm/Transforms/Scalar.h)
+#if defined (HAVE_LLVM_TRANSFORMS_SCALAR_GVN_H)
+#  include <llvm/Transforms/Scalar/GVN.h>
+#endif
+
 static bool Vdebug_jit = false;
 
 static bool Vjit_enable = false;
 
 static int Vjit_startcnt = 1000;
 
 static int Vjit_failcnt = 0;
 
